`storage`: fix `get_file_offset()` for truncated segments

WillemKauf opened the pull request:
This function could run into issues with the log `_start_offset` when making a `log_reader`.

Consider the case in which:
1. A prefix truncation intersects a `segment`, which bumps `_start_offset` but doesn't alter the `segment` or `segment_index` on file.
2. A call to `offset_range_size()` is issued.

Currently, we are either assuming the found entry from a `segment_index` has a `base_offset >= _start_offset` AND we're assuming that `segment` passed has `s->offsets().get_base_offset() >= _start_offset`, neither of which may be true due to the prefix truncation. This can lead to a thrown exception in the `log_reader`: `std::runtime_error: Reader cannot read
before start`.

Fix the problem by using a `single_segment_reader` (which reads the physical batches on disk, not subject to the same constraints of `log` state) instead of the `log_reader` to avoid the problems due to having to read before the new `_start_offset`.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [X] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

WillemKauf committed:
`storage`: add `single_segment_reader`

Meant for reading batches from a single `segment`.

This does not consider any other `log` state- meaning, the `log`
could have been prefix truncated and batches read from the `segment`
could be ahead of the new `_start_offset`.

This class should only be used when it is necessary to consider the
batches still physically stored in a `segment` on disk, e.g for file
position calculations. For all other applications, the `log_reader`
should be used.

WillemKauf committed:
`storage`: use `single_segment_reader` in `get_file_offset()`

This function could run into issues with the log `_start_offset` when
making a `log_reader`.

Consider the case in which:
1. A prefix truncation intersects a `segment`, which bumps `_start_offset`
   but doesn't alter the `segment` or `segment_index` on file.
2. A call to `offset_range_size()` is issued.

Currently, we are either assuming the found entry from a `segment_index`
has a `base_offset >= _start_offset` AND we're assuming that `segment`
passed has `s->offsets().get_base_offset() >= _start_offset`, neither of
which may be true due to the prefix truncation. This can lead to a thrown
exception in the `log_reader`: `std::runtime_error: Reader cannot read
before start`.

Fix the problem by using a `single_segment_reader` (which reads the physical
batches on disk, not subject to the same constraints of `log` state) instead
of the `log_reader` to avoid the problem of having to read before the new
`_start_offset`.

WillemKauf commented on src/v/storage/log_reader.h:
```diff
@@ -299,4 +300,51 @@ timequery_result batch_timequery(
   model::timestamp t,
   model::offset max_offset);
 
+class single_segment_reader final : public model::record_batch_reader::impl {
```
feels heavy-handed to add this new `impl` class. But we are concerned with the `segment`'s physical representation on disk in `get_file_offset()`, so we cannot be subject to the same constraints as the `log_reader`.

Lazin replied:
This is why `log_segment_batch_reader` is for. The impl feels redundant because it just adds additional nesting.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d875f677-1905-44a9-88dd-a81f6a50cfa8-->
<!--TEST RESULTS-->
<!--de519ab0-2703-487c-97b4-28ec6a24af36-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65588">build#65588</a></summary>


| test_id                                                                                                                                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.controller_log_limiting_test.ControllerLogLimitPartitionBalancerTests.test_partition_balancer_with_limits                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65588#0196a78a-9568-4b67-8b3c-3e3ecf67802a | FLAKY         | 20/21    |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65588#0196a783-ed3d-4a14-882e-25c1fdc9697c | FAIL          | 0/1      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65701">build#65701</a></summary>


| test_class                     | test_method                              | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------------|:-----------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeCustomPartitioningTest | test_spec_evolution_rpcn                 | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1}                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65701#0196b0a0-4631-4ed0-802a-8c957949e4d3 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| PartitionBalancerTest          | test_transfer_controller_leadership      |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65701#0196b08c-4cb6-4002-ac94-409cc9a5e0b2 | FLAKY         | 14/21    | upstream reliability is '82.80871670702179'. current run reliability is '66.66666666666666'. drift is 16.14205 and the allowed drift is set to 50. The test should PASS |
| PartitionBalancerTest          | test_unavailable_nodes                   |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65701#0196b08c-4cb3-40de-889f-3e534db6e374 | FAIL          | 0/1      |                                                                                                                                                                         |
| RandomNodeOperationsTest       | test_node_operations                     | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65701#0196b08c-4cb5-4b30-a267-d35990acbd10 | FLAKY         | 18/19    |                                                                                                                                                                         |
| RandomNodeOperationsTest       | test_node_operations                     | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65701#0196b08c-4cb3-40de-889f-3e534db6e374 | FLAKY         | 17/19    |                                                                                                                                                                         |
| TxAtomicProduceConsumeTest     | test_basic_tx_consumer_transform_produce | {"with_failures": true}                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65701#0196b0a0-4634-4a26-a4c0-925c49fd2b53 | FLAKY         | 20/21    | upstream reliability is '99.58847736625515'. current run reliability is '95.23809523809523'. drift is 4.35038 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

nvartolomei reviewed (commented):
lgtm

Few nits:
- Please add a doc/comment to the class. E.g. Copy the one in the git commit message because it is good.
- Consider making this class "private". It shouldn't leak above disk_log_impl. Maybe even move it entirely to `disk_log_impl.cc` or to a new cc file and export only a "factory" function. I.e. in a similar way `make_memory_record_batch_reader` is implemented. They are all pimpls anyway.

nvartolomei commented:
While you are at it, the other `offset_range_size` doesn't check that being and last are contained within the log after acquiring locks, only before. Can you double check that and the second check after locks are acquired to exclude a race condition? 🙇

WillemKauf committed:
`storage`: add utility functions for offset ranges

Including `log_contains_offset()` and `log_contains_offset_range()`,
for determining if a given offset `o` or an offset range `[first, last]`
are bounded within the log's offsets `[start_offset, committed_offset]`
respectively.

WillemKauf committed:
`storage`: use new utility functions in `get_offset_range()`

No functional changes.

WillemKauf committed:
`storage`: recheck `log` offsets after acquiring locks in `get_offset_range()`

To properly check whether we have raced with a prefix truncation or not,
re-check that the queried offset/offset range are still found within
the bounds of the log offsets.

Lazin commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -2206,12 +2207,16 @@ ss::future<size_t> disk_log_impl::get_file_offset(
 
-    auto reader = co_await make_reader(reader_cfg);
+    auto reader = model::make_record_batch_reader<single_segment_reader>(
+      s, co_await s->read_lock(), reader_cfg, *_probe);
```
isn't it equivalent to the `log_segment_batch_reader` with read lock stored on the stack? the reader is not passed anywhere and consumed on line 2222

WillemKauf replied:
`log_segment_batch_reader` is not a derived class of `model::record_batch_reader::impl`, so if we want to stick with the semantics of readers/consumers with the `batch_size_accumulator`, we need this implementation defined here. Otherwise, we would have to break the batch iterating logic out into `get_file_offset()`, i.e something like:


```diff
ss::future<size_t> disk_log_impl::get_file_offset(
  ss::lw_shared_ptr<segment> s,
  std::optional<segment_index::entry> maybe_index_entry,
  model::offset target,
  boundary_type boundary,
  ss::io_priority_class priority) {
    auto index_entry = maybe_index_entry.value_or(segment_index::entry{
      .offset = s->offsets().get_base_offset(),
      .filepos = 0,
    });
    size_t size_bytes{index_entry.filepos};
    details::batch_size_accumulator acc{
      .result_size_bytes = &size_bytes,
      .target = target,
      .boundary = boundary,
    };

    auto reader_start_offset = index_entry.offset;

    storage::log_reader_config reader_cfg(
      reader_start_offset, target, priority);

    reader_cfg.skip_batch_cache = true;
    reader_cfg.skip_readers_cache = true;

+    auto lock = co_await s->read_lock();
+    log_segment_batch_reader rdr = log_segment_batch_reader(
+      s, reader_cfg, *_probe);
+    auto batches = co_await rdr.read_some(model::no_timeout);
+    while (batches.has_value() && !batches.empty()) {
+        bool stop = false;
+        for (auto& batch : batches) {
+            stop = co_await acc(batch);
+            if (stop) {
+                break;
+            }
+        }
+        if (stop) {
+            break;
+        }
+        batches = co_await rdr.read_some(model::no_timeout);
+    }

    co_return size_bytes;
```

Let me know if you have a strong opinion here (leave the new impl class in, or simply write the iterating logic into `get_file_offset()`)

WillemKauf commented:
Force push to:
* Move `single_segment_reader` to `disk_log_impl.cc`
* Add new commit which adds utility functions `disk_log_impl::log_contains_offset()` and `disk_log_impl::log_contains_offset_range()`
* Add new commit which uses the newly added utility functions in `get_offset_ranges()` overloads
* Add new commit which re-checks that offset arguments passed to `get_offset_ranges()` are still bounded by the log after acquiring segment locks to prevent issues with concurrent prefix truncation.

WillemKauf commented:
@Lazin let me know if this looks good to you.
